src/target/riscv/riscv.c

```diff
@@ -643,6 +643,73 @@ int riscv_remove_watchpoint(struct target *target,
 	return ERROR_OK;
 }
 
+int riscv_hit_watchpoint(struct target *target, struct watchpoint **hit_watchpoint)
+{
+	struct watchpoint *wp = target->watchpoints;
+
+	LOG_DEBUG("Current hartid = %d", riscv_current_hartid(target));
+
```

Comment:
Instead of disassembling the instruction that we think caused the trigger, we should first check the hit bit of each watchpoint. This was added to the spec relatively recently (so I don't think it's actually implemented yet) but makes solving this problem simpler and more reliable. At a minimum we need a TODO comment here mentioning the bit.